metrics/scaling/k8s_scale_net.sh

```diff
@@ -0,0 +1,287 @@
+                            "time_to_pod_net": ${net_time},
+                            "Units" : "ms"
+                    }
+EOF
+)"
+    metrics_json_add_array_fragment "$pods_json"
+
+    local launch_json="$(cat << EOF
+                        "launch_time": {
+                                "Result": $launch_time_ms,
+                                "Units" : "ms"
+                            }
+EOF
+)"
+    metrics_json_add_array_fragment "$launch_json"
+
+    info "launch [$launch_time_ms]"
+
+    metrics_json_close_array_element
+}
+
+init() {
+    info "Initialising"
+
+    local cmds=("bc" "jq")
+    check_cmds "${cmds[@]}"
+
+    info "Checking Kubernetes accessible"
+    local worked=$( kubectl get nodes > /dev/null 2>&1 && echo $? || echo $? )
+    if [ "$worked" != 0 ]; then
+        die "kubectl failed to get nodes"
+    fi
+
+    info $(get_num_nodes) "Kubernetes nodes in 'Ready' state found"
+
+    k8s_api_init
+
+     # And now we can set up our results storage then...
+     metrics_json_init "k8s"
+     save_config
+}
+
+save_config() {
+    metrics_json_start_array
+
+    local json="$(cat << EOF
+    {
+        "testname": "${TEST_NAME}",
+        "NUM_DEPLOYMENTS": ${NUM_DEPLOYMENTS},
+        "STEP": ${STEP},
+        "wait_time": ${wait_time},
+        "delete_wait_time": ${delete_wait_time},
+        "settle_time": ${settle_time}
+    }
+EOF
+)"
+	metrics_json_add_array_element "$json"
+	metrics_json_end_array "Config"
+}
+
+run() {
+	info "Running test"
+	local header_post="Content-Type: application/json"
+	local base_curl=${API_ADDRESS}:${API_PORT}/apis/apps/v1/namespaces/default/deployments
+
+	trap cleanup EXIT QUIT KILL
+
+	metrics_json_start_array
+
+	for reqs in $(seq ${STEP} ${STEP} ${NUM_DEPLOYMENTS}); do
+		local deployment="${name_base_depl}${reqs}"
+		info "Testing replicas ${reqs} of ${NUM_DEPLOYMENTS}"
+		# Generate the next yaml file
+
+		local runtime_command
+		if [ -n "$RUNTIME" ]; then
+			runtime_command="s|@RUNTIMECLASS@|${RUNTIME}|g"
+		else
+			runtime_command="/@RUNTIMECLASS@/d"
+		fi
+
+		local input_template
+		local generated_file
+		if [ "$use_api" != "no" ]; then
+			input_template=$input_json
+			generated_file=$generated_json
+		else
+			input_template=$input_yaml
+			generated_file=$generated_yaml
+		fi
+
+		sed -e $runtime_command \
+			-e "s|@DEPLOYMENT@|${deployment}|g" \
+			-e "s|@LABEL@|${LABEL}|g" \
+			-e "s|@LABELVALUE@|${LABELVALUE}|g" \
+			-e "s|@GRACE@|${grace}|g" \
+			< ${input_template} > ${generated_file}
+
+		info "Applying changes"
+		local start_time=$(date +%s%N)
+
+		if [ "$use_api" != "no" ]; then
+			curl -s ${base_curl} -XPOST -H "${header_post}" -d@${generated_file} > /dev/null
+		else
+			kubectl apply -f ${generated_file}
+		fi
+
+		kubectl rollout status --timeout=${wait_time}s deployment/${deployment}
+		kubectl expose deployment $deployment
+
+		set +e
+		while [ $? -ne 0 ]; do
+			kubectl get services $deployment -n default --no-headers=true
```

Comment:
Looking at this again - that is a hard non-timeout spin loop - we can't have that in the tests, it could lock up forever.
Can you use the `--timeout` option on the `get services` with `$wait_time`, and not do the spin loop - just let the `kubectl` fail if it does not find the service within the timeout period?

Reply:
OK @marcemq , so let's solve both the inline use of `+e`, and the hard spin loop....

> Note: my main objection to the `+e` is, it is just bad habit to open code that inline - in this specific case, as it is in a top level piece of code, it would probably be 'safe', but, if this were for instance in a piece of 'library' code, the problem is you do not know if the calling code has `e` set or not - so you will struggle to set its value back to the original...

First, the easiest way to get around the `set +e` mess would I think to be put it into a 'subshell', like:

```patch
@@ -153,11 +153,13 @@ run() {
                kubectl rollout status --timeout=${wait_time}s deployment/${deployment}
                kubectl expose deployment $deployment

-               set +e
-               while [ $? -ne 0 ]; do
-                       kubectl get services $deployment -n default --no-headers=true
-               done
-               set -e
+               #Execute in subshell so we can drop the error check...
+               (
+                       set +e
+                       while [ $? -ne 0 ]; do
+                               kubectl get services $deployment -n default --no-headers=true
+                       done
+               )

                IP=$(kubectl get services $deployment -n default --no-headers=true | awk '{printf $3}')
                end_net=$(date +%s%N)
```

Having said that.... I went looking for some example code of how to do a timeout loop, as we need to not have the hard spin 'death loop' scenario. If we grab the [`waitForProcess` code from the Kata project](https://github.com/kata-containers/tests/blob/master/.ci/lib.sh#L160-L173):

```bash
function waitForProcess(){
        wait_time="$1"
        sleep_time="$2"
        cmd="$3"
        while [ "$wait_time" -gt 0 ]; do
                if eval "$cmd"; then
                        return 0
                else
                        sleep "$sleep_time"
                        wait_time=$((wait_time-sleep_time))
                fi
        done
        return 1
}
```

We can see that looks to use an `eval` to get around the error condition.

let's go look for a relevant [use example from kata](https://github.com/kata-containers/tests/blob/master/integration/kubernetes/k8s-expose-ip.bats#L32-L34)

```bash
# Check pods are running
cmd="kubectl get pods -o jsonpath='{.items[*].status.phase}' | grep Running"
waitForProcess "$wait_time" "$sleep_time" "$cmd"
```

My recommendation then is you add the function `waitForProcess` to our library code, and use it to do the service checks, which should I think cure both the `set +e` and the hard spin loop timeout problems.

Reply:
Done.

And the the values used for [`wait_time_proc` and `sleep_time_proc`](https://github.com/kata-containers/tests/blob/master/integration/kubernetes/k8s-expose-ip.bats#L32-L34) used are the ones used in kata project too.